### ISSUE TITLE
[5.x]: Eager loading of multiple added fields return the same entries for a related field

### ISSUE THREAD

**Issue by naboo** [craftcms/cms#16191]
### Description

I have an entry in my PRODUCTS section which has a ENTRIES field added to it to load an entry from my COLLECTION section. My collection entry then has two ENTRIES field added to it (using the same field) but with two different handles.

When running this code to fetch the related eager-loaded entries from the collection entry they return the same entries even if one of the fields doesn't even have entries selected.

```
{% set collection = entry.collectionEntry.eagerly().one() %}
{% set relatedEntries = collection.relatedEntries.eagerly().all() %}
{% set moreRelatedEntries = collection.moreRelatedEntries.eagerly().all() %}
```

... so "relatedEntries" and "moreRelatedEntries" pull the same entries - but only entries are selected for field "moreRelatedEntries". They are both the same field added to the section COLLECTION multiple times.

Related to: https://github.com/craftcms/cms/issues/15890

### Expected behavior

Only the actual entries selected in the related ENTRIES field type should return for each specific field.

### Craft CMS version

5.5.x

### PHP version

8.3

### Operating system and version

Ubuntu 22.x

### Database type and version

MySQL 8.0+

### Installed plugins and versions

-

**Comment by brandonkelly** [comment]
Thanks for reporting that! We just released Craft 5.5.10 with a fix.

**Comment by dgsiegel** [comment]
@brandonkelly that doesn't quite seem to work with the same field inside a Matrix field in Craft 5.5.10:

```
{% set items = block.someEntries.all() %}
{% for item in items %}
  {{ item.image.eagerly().one().url }}
{% endfor %}

{% for item in items %}
  {% set images = item.image.eagerly().all() %}
  {% for image in images %}
    {{ image.url }}
  {% endfor %}
{% endfor %}
```

This will only print the first asset, even though that field contains multiple.

**Comment by brandonkelly** [comment]
@dgsiegel In that example, the `image` field is going to get eager-loaded with `limit=1` because you are calling `.one()` the first time it’s executed.

If you want to eager-load the same field multiple times with different params/limit, you can pass a custom alias into `eagerly()` for one or both of the queries:

```twig
{% set items = block.someEntries.all() %}
{% for item in items %}
  {{ item.image.eagerly('image-one').one().url }}
{% endfor %}

{% for item in items %}
  {% set images = item.image.eagerly('image-all').all() %}
  {% for image in images %}
    {{ image.url }}
  {% endfor %}
{% endfor %}
```

Alternatively, you could improve performance by only eager-loading _all_ of the images, and then outputting the first one from that array with `|first` rather than `.one()`:

```twig
{% set items = block.someEntries.all() %}

{% for item in items %}
  {% set image = item.image.eagerly().all()|first %}
  {{ image.url ?? '' }}
{% endfor %}

{% for item in items %}
  {% set images = item.image.eagerly().all() %}
  {% for image in images %}
    {{ image.url }}
  {% endfor %}
{% endfor %}
```

**Comment by dgsiegel** [comment]
@brandonkelly thanks for clearing this up! It might make sense to add the alias to the documentation, as I haven't seen it before.

We've taken the second approach already. Will the duplicate `item.image.eagerly().all()` impact performance or is Craft clever enough to only execute in once? Thanks!

**Comment by brandonkelly** [comment]
@dgsiegel Yep, in that scenario the `image` relations will only get loaded once, for the first query.
